python/ray/tune/impl/tuner_internal.py

```diff
@@ -0,0 +1,209 @@
+                        [isinstance(v_item, Dataset) for v_item in v]
+                    ):
+                        raise TuneError(
+                            "Wrongly formatted dataset param passed in Tune!"
+                        )
+                    if len(v) > 0 and isinstance(v[0], Dataset):
+                        dataset_dict[k] = [v_item.fully_executed() for v_item in v]
+                else:
+                    raise TuneError("Unexpected dataset param passed in.")
+
+        if _DATASETS in self.param_space:
+            ds = self.param_space[_DATASETS]
+            if isinstance(ds, Dataset):
+                self.param_space[_DATASETS] = ds.fully_executed()
+            elif isinstance(ds, dict):
+                _helper(ds)
+            else:
+                # We shouldn't be expecting anything here.
+                raise TuneError("Unexpected dataset param passed in.")
+
+    def _get_or_create_experiment_checkpoint_dir(self, local_dir: Optional[str]):
+        """Get experiment checkpoint dir before actually running experiment."""
+        path = Experiment.get_experiment_checkpoint_dir(
+            self._convert_trainable(self.trainable), local_dir, self.name
+        )
+        if not os.path.exists(path):
+            os.makedirs(path)
+        return path
+
+    # This has to be done through a function signature (@property won't do).
+    def experiment_checkpoint_dir(self):
+        return self.experiment_checkpoint_dir
+
+    @staticmethod
+    def _convert_trainable(trainable: Any):
+        if isinstance(trainable, ConvertibleToTrainable):
+            trainable = trainable.as_trainable()
+        else:
+            trainable = trainable
+        return trainable
+
+    def fit(self):
+        trainable = self._convert_trainable(self.trainable)
+        if not self.is_restored:
+            param_space = copy.deepcopy(self.param_space)
+            analysis = self._fit_internal(trainable, param_space)
+        else:
+            analysis = self._fit_resume(trainable)
+
+        return analysis
+
+    def _fit_internal(self, trainable, param_space):
+        """Fitting for a fresh Tuner."""
+        assert self.experiment_checkpoint_dir
+        analysis = run(
+            trainable,
+            config={**param_space},
+            name=self.name,
+            callbacks=self.callbacks,
+            _experiment_checkpoint_dir=self.experiment_checkpoint_dir,
+        )
+        return analysis
+
+    def _fit_resume(self, trainable):
+        """Fitting for a restored Tuner."""
+        assert self.experiment_checkpoint_dir
+        analysis = run(
+            trainable,
+            resume=True,
+            callbacks=self.callbacks,
+            _experiment_checkpoint_dir=self.experiment_checkpoint_dir,
+        )
+        return analysis
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state.pop("trainable", None)
+        state.pop("param_space", None)
```

Comment:
why pop these?

Reply:
trainable field is checkpointed separately. 
param_space is not needed for resuming a tune run, as the information is already saved in Searcher and trials.